packages/next/client/components/app-router.tsx

```diff
@@ -291,7 +292,10 @@ function Router({
     // __NA is used to identify if the history entry can be handled by the app-router.
     // __N is used to identify if the history entry can be handled by the old router.
     const historyState = { __NA: true, tree }
-    if (pushRef.pendingPush) {
+    if (
+      pushRef.pendingPush &&
+      createHrefFromUrl(new URL(window.location.href)) !== canonicalUrl
```

Comment:
Since window.location is a Location object that has all the same fields as URL, maybe we can just use that instead of turning it into a string, parsing it into a new object and then reassemble.

Reply:
Opened a PR here: https://github.com/vercel/next.js/pull/42888